src/main/java/com/powsybl/openloadflow/TargetVoltageCompatibilityChecker.java

```diff
@@ -0,0 +1,139 @@
+import com.powsybl.openloadflow.network.LfBus;
+import com.powsybl.openloadflow.network.LfNetwork;
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.apache.commons.lang3.tuple.Pair;
+import org.jgrapht.Graph;
+import org.jgrapht.graph.Pseudograph;
+import org.jgrapht.traverse.BreadthFirstIterator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * @author Geoffroy Jamgotchian {@literal <geoffroy.jamgotchian at rte-france.com>}
+ */
+public class TargetVoltageCompatibilityChecker {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TargetVoltageCompatibilityChecker.class);
+
+    private final LfNetwork network;
+
+    public TargetVoltageCompatibilityChecker(LfNetwork network) {
+        this.network = Objects.requireNonNull(network);
+    }
+
+    private static Graph<LfBus, LfBranch> createGraph(LfNetwork lfNetwork) {
+        Graph<LfBus, LfBranch> graph = new Pseudograph<>(LfBranch.class);
+        for (LfBranch branch : lfNetwork.getBranches()) {
+            LfBus bus1 = branch.getBus1();
+            LfBus bus2 = branch.getBus2();
+            if (bus1 != null && bus2 != null && !branch.isDisabled()) {
+                if (!graph.containsVertex(bus1)) {
+                    graph.addVertex(bus1);
+                }
+                if (!graph.containsVertex(bus2)) {
+                    graph.addVertex(bus2);
+                }
+                graph.addEdge(bus1, bus2, branch);
+            }
+        }
+        return graph;
+    }
+
+    private static Set<LfBus> exploreNeighbors(Graph<LfBus, LfBranch> graph, LfBus bus, int maxDepth) {
+        var it = new BreadthFirstIterator<>(graph, bus);
+        Set<LfBus> neighbors = new HashSet<>();
+        while (it.hasNext()) {
+            LfBus b = it.next();
+            int currentDepth = it.getDepth(b);
+            if (currentDepth > maxDepth) {
+                break;
+            }
+            if (b != bus) {
+                neighbors.add(b);
+            }
+        }
+        return neighbors;
+    }
+
+    public List<Pair<LfBus, LfBus>> check(TargetVoltageCompatibilityCheckerParameters parameters) {
+        Stopwatch stopwatch = Stopwatch.createStarted();
+
+        Graph<LfBus, LfBranch> graph = createGraph(network);
+
+        Set<LfBus> controlledBuses = network.getBuses().stream()
+                .filter(bus -> !bus.isDisabled()
+                        && bus.isVoltageControlled()
+                        && bus.getVoltageControls().stream().anyMatch(vc -> !vc.isDisabled()))
+                .collect(Collectors.toSet());
+
+        List<Pair<LfBus, LfBus>> incompatibleControlledBuses = new ArrayList<>();
+
+        var ySystem = AdmittanceEquationSystem.create(network, new VariableSet<>());
+        try (var y = AdmittanceMatrix.create(ySystem, parameters.getMatrixFactory())) {
+            for (LfBus controlledBus : controlledBuses) {
```

Comment:
Don't you compute the impedance between each pair of bus twice ? if you iterate on all buses then on all neighbours ?

Also, would it accerlerate things if there was an option to check a pair only if at least one of the two buses is remotely controlled ? [This is where we have convergence issues]

Reply:
Yes, we are checking potentially twice. It could be improved indeed.